pkg/reconciler/ingress/tracker.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ingress
+
+import (
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	"k8s.io/klog/v2"
+)
+
+// tracker is used to track the relationship between services and ingresses.
+// It is used to determine which ingresses are affected by a service change and
+// trigger a reconciliation of the affected ingresses.
+type tracker struct {
+	mu                sync.Mutex
+	trackedServices   map[string][]networkingv1.Ingress
+	ingressToServices map[string]map[string]struct{}
+}
+
+// newTracker creates a new tracker.
+func newTracker() *tracker {
+	return &tracker{
+		trackedServices:   make(map[string][]networkingv1.Ingress),
+		ingressToServices: make(map[string]map[string]struct{}),
+	}
+}
+
+// getIngress returns the list of ingresses that are related to a given service, or
+// an empty list, and a boolean, false, indicating if the list is empty.
+func (t *tracker) getIngress(service *corev1.Service) ([]networkingv1.Ingress, bool) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	s, ok := t.trackedServices[serviceToKey(service)]
+	return s, ok
+}
+
+// Adds a service to an ingress (key) to be tracked.
+func (t *tracker) add(ingress *networkingv1.Ingress, s *corev1.Service) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	klog.Infof("tracking service %q for ingress %q", s.Name, ingress.Name)
+	for _, ti := range t.trackedServices[serviceToKey(s)] {
+		if ingressToKey(&ti) == ingressToKey(ingress) {
+			return
+		}
+	}
+	t.trackedServices[serviceToKey(s)] = append(t.trackedServices[serviceToKey(s)], *ingress)
+	t.ingressToServices[ingressToKey(ingress)] = make(map[string]struct{})
```

Comment:
Can you ever have one ingress map to multiple services? If so, this appears to clear out the map each time this function is called?